web/telemetry/host_telemetry.go

```diff
@@ -0,0 +1,61 @@
+package telemetry
+
+import (
+	"errors"
+	"time"
+
+	"github.com/trento-project/trento/web/models"
+	"gorm.io/gorm"
+)
+
+type HostTelemetry struct {
+	InstallationID string    `json:"installation_id"`
+	AgentID        string    `json:"agent_id"`
+	SLESVersion    string    `json:"sles_version"`
+	CPUCount       int       `json:"cpu_count"`
+	SocketCount    int       `json:"socket_count"`
+	TotalMemoryMB  int       `json:"total_memory_mb"`
+	CloudProvider  string    `json:"cloud_provider"`
+	Time           time.Time `json:"time"`
+}
+
+type HostTelemetries []*HostTelemetry
+
+type HostTelemetryExtractor struct {
+	installationIdAwareExtractor
+	db *gorm.DB
+}
+
+func (ex *HostTelemetryExtractor) Extract() (interface{}, error) {
+	var collectedHostsTelemetry []models.HostTelemetry
+	var publishableHostTelemetries HostTelemetries
+
+	if err := ex.db.Find(&collectedHostsTelemetry).Error; err != nil {
+		return nil, err
+	}
+
+	if len(collectedHostsTelemetry) == 0 {
```

Comment:
Are we sure we want to treat this as an error?
We might have some initial installation without attaching any host (strange, but could happen). Would this affect somehow?

Reply:
Returning an error from an `Extractor` affects the following publishing step, which wouldn't be executed.
That's by design

Look a [this line](https://github.com/trento-project/trento/pull/487/files#diff-7c76a17fd58e6a30285ef73042c129f98002c0edcfbd6fe8aed48274eddd67c5R55)
```
for telemetryName, extractor := range *e.telemetryRegistry {
	...
	extractedTelemetry, err := extractor.Extract()
	if err != nil {
		log.Errorf("Error while extracting telemetry %s: %s", telemetryName, err)
		continue
	}
        ... then publish
}
```
Does this answer and make sense? Am I overlooking anything?

Reply:
My question is more about error treatment. Not having any record is not an error for me. It won't publish anything and that's all.
This is how I see it.
Anyway, so far it wouldn't change in anything the behaviour

Reply:
Okay, I see what you mean and it also makes sense.
Thanks for clarifying.

Returning an error is a way I found to communicate that for this extractor we do not need to actually publish anything.
Clearly there are other options.

Feel free to share some alternative options you see valid, so we can keep them into account next iterations.

Reply:
I would personally do:
- If we don't have data, we send an empty list without an error
- In the `extractAndPublishFn` function, after checking if there was any error, I would check for the length of the received data. If 0, we continue the loop. Otherwise, we do what we do now.

For me it is clearer.
Anyway, as commented, both options are good enough.

Reply:
I see, that'd be valid.

However, `Extract` returns an `interface{}` so we don't really know what's in there, so first issue I see is that we are assuming that `Extract` method returns something I cal call `len` on, which might not be always the case :thinking: 

An hybrid approach would be:
- expecting that `Extract` still returns an `interface{}`, yet by design we define that one of the possible values for that return could be something of type `SkipPublishing`
- make the extractors return a `SkipPublishing` thing when they want to communicate that nothing should happen at publishing level for that extractor
- check in the engine the returned type from `Extract` method

```
type SkipPublishing struct{}

extractedTelemetry, err := extractor.Extract()
if _, ok := extractedTelemetry.(SkipPublishing); ok {
	continue
}

```
I don't dislike this either, but it seems too much.

Just to say I am open to change :wink: 
Will keep this under the radar. Thanks.